sklearn/utils/validation.py

```diff
@@ -892,9 +903,13 @@ def check_is_fitted(estimator, msg=None):
     if not hasattr(estimator, 'fit'):
         raise TypeError("%s is not an estimator instance." % (estimator))
 
-    attrs = [v for v in vars(estimator)
-             if (v.endswith("_") or v.startswith("_"))
-             and not v.startswith("__")]
+    if attributes is not None:
+        if not isinstance(attributes, (list, tuple)):
+            attributes = [attributes]
+        attrs = all([hasattr(estimator, attr) for attr in attributes])
+    else:
+        attrs = [v for v in vars(estimator)
+                 if v.endswith("_") and not v.startswith("__")]
```

Comment:
why do we keep ``not v.startswith("__")``?

Reply:
Because all dunder attributes end with `_` and are not fit parameters.

Reply:
And more generally, anything that starts with `__` is weird / reserved in Python so scikit-learn attributes should never start with `__`.

Reply:
> And more generally, anything that starts with __ is weird / reserved in Python so scikit-learn attributes should never start with __.

You're right, but perhaps it's better to remove, because we do not forbid users to create attributes start with ``__``.

Reply:
If we remove it we will have false positives.

Reply:
> If we remove it we will have false positives.

Could you please provide more details?

Reply:
@qinhanmin2014 is right that we don't need to worry about magic attributes, since we are using `vars` and not `dir`, and these are all part of the class's `vars`, not part of the instance's `vars`.... at least as far as I know. While this may be true for python attributes (see `__len__` below), I'm not sure that it's explicitly the case for dunder magic defined elsewhere, e.g. `__array__`.

```
In [1]: class MyFunkyThing:
   ...:     def __init__(self):
   ...:         def yuck():
   ...:             return np.arange(5)
   ...:         self.__array__ = yuck
   ...:

In [2]: np.asarray(MyFunkyThing())
Out[2]: array([0, 1, 2, 3, 4])

In [3]: vars(MyFunkyThing())
Out[3]: {'__array__': <function __main__.MyFunkyThing.__init__.<locals>.yuck()>}

In [4]:
   ...: class MyFunkyThing:
   ...:     def __init__(self):
   ...:         def yuck():
   ...:             return 5
   ...:         self.__len__ = yuck
   ...:

In [5]: len(MyFunkyThing())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-584e51b5dabc> in <module>
----> 1 len(MyFunkyThing())

TypeError: object of type 'MyFunkyThing' has no len()
```

This is really getting deep into technicalities... in practice there is little risk of not excluding `__.*__` here.

Reply:
But, I should say, I'd rather exclude magic dunder names here.

Reply:
perhaps a clearer solution is ``if v.endswith("_") and not v.endswith("__")``? @jnothman

Reply:
+1 to keep excluding based on `v.startswith("__")`. I have seen users use private variables in `__init__` in their estimators, it's not good, but we shouldn't count those as fit attributes either.

Reply:
> I have seen users use private variables in __init__ in their estimators, it's not good, but we shouldn't count those as fit attributes either.

but @rth we're talking about double underscore, not single underscore (private variables).

Reply:
Anyway I guess this is not so important.

Reply:
> but @rth we're talking about double underscore, not single underscore (private variables).

I mean technically private variables as in `self.__a`, not very frequent but I have seen it. Anyway, yes it's not too critical either way. I just wanted to merge this quickly to fix CI on scikit-learn-extra. We can always come back on these detail at later time if needed.